_posts/2020-03-13-password-protected-efk-stack-on-k8s.md

```diff
@@ -0,0 +1,423 @@
+
+I always prefer minikube for such experimental works, which makes easy to run k8s locally. Please note that minikube uses 1GB of memory and 2 CPU cores by default and these values are not sufficient for our EFK stack. You should start/configure minikube with the following command:
+
+``` yml
+$ minikube start --memory 8192 --cpus 4
+```
+
+If you have installed minikube successfully and run it with above configuration, you're about halfway through the EFK stack :)
+
+---
+
+To get started, create a namespace inside minikube, to do that you can simply run the following command:
+
+``` yml
+$ kubectl create ns kube-logging
+```
+
+A minor detail, you're going to find a file inside each module named as _kustomization.yaml_, [kustomize]https://kustomize.io is another facilitator which lets us customize raw, template-free YAML files for multiple purposes, leaving the original YAML untouched and usable as is.
+
+***
+
+## Elasticsearch
+
+Let's get started to walk through the details of Elasticsearch which consists of a Statefulset, service, persistent volume and ConfigMap. You can always have a quick look to get the component summary from its _kustomization.yaml_ file.
+Firstly, we will claim a persistent volume from k8s for our Elasticsearch Statefulset which will be used to store the data collected by Fluentd.
+
+Firstly, we will claim a persistent volume from k8s for our Elasticsearch Statefulset which will be used to store the data collected by Fluentd.
+
+``` yml
+resources:
+- statefulset.yaml
+- service.yaml
+- configmap.yaml
+- pvc.yaml
+```
+
+Secondly, we will deploy a service to expose our Elasticsearch application running on a set of pods as a network service, and all components running in our cluster would be able to reach Elasticsearch through port 9200.
+
+``` yml
+apiVersion: v1
+kind: Service
+metadata:
+  name: elasticsearch
+  labels:
+    component: elasticsearch
+spec:
+  type: NodePort
+  selector:
+    component: elasticsearch
+  ports:
+  - port: 9200
+    targetPort: 9200
+```
+
+Then, our Elasticsearch application will be deployed as Statefulset with the following configuration. More importantly, we will create a user immediately after the container started, which will be used in Kibana and Fluentd to access Elasticsearch.
+
+``` yml
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: elasticsearch
+spec:
+  serviceName: elasticsearch
+  replicas: 1
+  selector:
+    matchLabels:
+      component: elasticsearch
+  template:
+    metadata:
+      labels:
+        component: elasticsearch
+    spec:
+      containers:
+      - name: elasticsearch
+        image: docker.elastic.co/elasticsearch/elasticsearch:6.8.6
+        imagePullPolicy: IfNotPresent
+        # Add post start lifecycle to add elasticsearch user
+        lifecycle:
+          postStart:
+           exec:
+            command: ["/bin/sh", "-c", "/usr/share/elasticsearch/bin/elasticsearch-users useradd esUser -p esPassword -r superuser"]
```

Comment:
Is there any better way to setup username and password for ES without embedding it as a part of stateful-set description? In this case, once anyone queries pod description, he sees the password.
The rest of tutorial uses config-maps, that can be converted to secrets and you can apply rbac and sealed-secrets on top of it. With command-exec, there is no option at all (or I do not know about it?)

If providing the proper solution in the same post is too much, maybe put a disclaimer at the beginning, which gives directions how to enhance the setup in _production_ case?

Reply:
I think instead of having this process in statefulset description, it would be better to create user and password by a built-in elasticsearch command https://www.elastic.co/guide/en/elasticsearch/reference/current/setup-passwords.html

I updated the article accordingly.